### PR TITLE
add Tensor Operation and Reduction cudnn API

### PR DESCRIPTION
**Pull request by kashif**

### CODE REVIEW

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by kashif** [comment]
@okuta I can also add the reduction operations API to this PR if you like?

**Comment by okuta** [comment]
OK

File: cupy/cuda/cudnn.pyx [hunk 1]
```diff
@@ -640,6 +657,48 @@ cpdef addTensor_v3(size_t handle, size_t alpha, size_t bDesc,
...
+    return opTensorOp, opTensorCompType, opTensorNanOpt
+
+cpdef destroyOpTensorDescriptor(size_t opTensorDesc):
+    status = cudnnDestroyOpTensorDescriptor(<OpTensorDescriptor>opTensorDesc)
+    check_status(status)
+
+cpdef opTensor(size_t handle, size_t opTensorDesc, size_t alpha1,
+               size_t aDesc, size_t A, size_t alpha2, size_t bDesc,
+               size_t B, size_t beta, size_t cDesc, size_t C):
+    status = cudnnOpTensor(
```
Comment (kmaehashi) [inline cupy/cuda/cudnn.pyx]: Do we need `setStream` (and thus `with nogil`) here?

Comment (kashif) [reply cupy/cuda/cudnn.pyx]: yes we do. Adding that now.

**Comment by kashif** [comment]
@okuta happy for another review

**Comment by okuta** [comment]
Thanks

jenkins, test this please.

File: cupy/cuda/cudnn.pyx [hunk 2]
```diff
@@ -640,6 +693,149 @@ cpdef addTensor_v3(size_t handle, size_t alpha, size_t bDesc,
     check_status(status)
 
 
+###############################################################################
+# Tensor operations
+###############################################################################
+
+cpdef size_t createOpTensorDescriptor() except *:
```
Comment (okuta) [inline cupy/cuda/cudnn.pyx]: Please use `? 0` instead of `*`.

File: cupy/cuda/cudnn.pyx [hunk 3]
```diff
@@ -640,6 +693,149 @@ cpdef addTensor_v3(size_t handle, size_t alpha, size_t bDesc,
...
+            <TensorDescriptor>bDesc, <void*>B, <void*>beta,
+            <TensorDescriptor>cDesc, <void*>C)
+    check_status(status)
+
+
+###############################################################################
+# Tensor reductions
+###############################################################################
+
+cpdef size_t createReduceTensorDescriptor() except *:
```
Comment (okuta) [inline cupy/cuda/cudnn.pyx]: Please use `? 0` instead of `*`.

File: cupy/cuda/cudnn.pyx [hunk 4]
```diff
@@ -640,6 +693,149 @@ cpdef addTensor_v3(size_t handle, size_t alpha, size_t bDesc,
...
+
+
+cpdef destroyReduceTensorDescriptor(size_t reduceTensorDesc):
+    status = cudnnDestroyReduceTensorDescriptor(
+        <ReduceTensorDescriptor>reduceTensorDesc)
+    check_status(status)
+
+
+cpdef size_t getReductionIndicesSize(size_t handle, size_t reduceTensorDesc,
+                                     size_t aDesc, size_t cDesc) except *:
```
Comment (okuta) [inline cupy/cuda/cudnn.pyx]: Please use `? 0` instead of `*`.

File: cupy/cuda/cudnn.pyx [hunk 5]
```diff
@@ -640,6 +693,149 @@ cpdef addTensor_v3(size_t handle, size_t alpha, size_t bDesc,
...
+    cdef size_t sizeInBytes
+    status = cudnnGetReductionIndicesSize(
+        <Handle>handle, <ReduceTensorDescriptor>reduceTensorDesc,
+        <TensorDescriptor>aDesc, <TensorDescriptor>cDesc, &sizeInBytes)
+    check_status(status)
+    return sizeInBytes
+
+
+cpdef size_t getReductionWorkspaceSize(size_t handle, size_t reduceTensorDesc,
+                                       size_t aDesc, size_t cDesc) except *:
```
Comment (okuta) [inline cupy/cuda/cudnn.pyx]: Please use `? 0` instead of `*`.

**Comment by kashif** [comment]
@okuta doesnt the `except? 0` imply that whenever an exception occurs it will return a `0` which is actually `CUDNN_STATUS_SUCCESS` and whenever these methods return `CUDNN_STATUS_SUCCESS` an exception will be assumed?

**Comment by okuta** [comment]
> And, please check [this PR](https://github.com/kashif/cupy/pull/1)
> Cloud you merge or fix.

**Comment by okuta** [comment]
Thanks.
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Comment by kashif** [comment]
thanks @okuta 🙇

**Pull request merged**